provider/gce/environ_network.go

```diff
@@ -1,40 +1,294 @@
+		}
+		if subnetIds.Include(subnet.Name) {
+			results = append(results, makeSubnetInfo(
+				network.Id(subnet.Name),
+				network.Id(networkId),
+				subnet.IpCidrRange,
+				zones,
+			))
+		}
+	}
+	return results, nil
+}
+
+func (e *environ) getInstanceSubnets(inst instance.Id, subnetIds IncludeSet, zones []string) ([]network.SubnetInfo, error) {
+	ifaces, err := e.NetworkInterfaces(inst)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var results []network.SubnetInfo
+	for _, iface := range ifaces {
+		if subnetIds.Include(string(iface.ProviderSubnetId)) {
+			results = append(results, makeSubnetInfo(
+				iface.ProviderSubnetId,
+				iface.ProviderNetworkId,
+				iface.CIDR,
+				zones,
+			))
+		}
+	}
+	return results, nil
+}
+
+// NetworkInterfaces implements environs.NetworkingEnviron.
+func (e *environ) NetworkInterfaces(instId instance.Id) ([]network.InterfaceInfo, error) {
+	insts, err := e.Instances([]instance.Id{instId})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	envInst, ok := insts[0].(*environInstance)
+	if !ok {
+		// This shouldn't happen.
+		return nil, errors.Errorf("couldn't extract google instance for %q", instId)
+	}
+	googleInst := envInst.base
+	ifaces := googleInst.NetworkInterfaces()
+
+	subnetURLs := make([]string, len(ifaces))
+	for i, iface := range ifaces {
+		subnetURLs[i] = iface.Subnetwork
+	}
+	subnets, err := e.subnetsByURL(subnetURLs...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// We know there'll be a subnet for each url requested, otherwise
+	// there would have been an error.
+
+	var results []network.InterfaceInfo
+	for i, iface := range ifaces {
+		subnet, ok := subnets[iface.Subnetwork]
+		if !ok {
+			// Should never happen.
+			return nil, errors.Errorf("no subnet %q found for instance %q", iface.Subnetwork, instId)
+		}
+
+		results = append(results, network.InterfaceInfo{
+			DeviceIndex: i,
+			CIDR:        subnet.CIDR,
+			// The network interface has no id in GCE so it's
+			// identified by the machine's id + its position.
+			ProviderId:        network.Id(fmt.Sprintf("%s/%d", instId, i)),
+			ProviderSubnetId:  subnet.ProviderId,
+			ProviderNetworkId: subnet.ProviderNetworkId,
+			AvailabilityZones: subnet.AvailabilityZones,
+			InterfaceName:     iface.Name,
+			Address:           network.NewScopedAddress(iface.NetworkIP, network.ScopeCloudLocal),
+			InterfaceType:     network.EthernetInterface,
+			Disabled:          false,
+			NoAutoStart:       false,
+			ConfigType:        network.ConfigDHCP,
+		})
+	}
+	return results, nil
+}
+
+func (e *environ) subnetsByURL(urls ...string) (map[string]network.SubnetInfo, error) {
+	// In GCE all the subnets are in all AZs.
+	zones, err := e.zoneNames()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	networks, err := e.networkIdsByURL()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// We don't want to get all subnets if no urls are passed.
```

Comment:
It was meant to explain why I'm not calling something that could return an includeAny, but you're right, doing an early return here makes more sense.